WordPress/gutenberg
# Global styles revisions: add pagination

Resolves https://github.com/WordPress/gutenberg/issues/53621

Part of:

- https://github.com/WordPress/gutenberg/issues/55776

## What?

Now that we have a [revisions API in Core Data](https://github.com/WordPress/gutenberg/pull/54046), let's add paginated results to global styles revisions!

Uploading 2023-12-19 15.40.44.mp4…

## Why?

Good question. Because currently, there is a hard limit of 100 results for global styles revisions coming straight from the  REST response.

Paginated results not only reduces the amount of space revisions take up in the sidebar, but allows users to access any number of revisions.

## How?
- Passing query parameters with `per_page` and `page` values to `useGlobalStylesRevisions()` to get paginated results.
- Adding a pagination component at the footer of the revisions panel.

## Testing Instructions

Requires a block theme like 2024.

In the Site Editor:

1. Check that the pagination navigation does not appear with < 10 revisions. A "revision" is classed as a saved revision record in the database (not any unsaved changes or "Reset" items in the revisions list).
2. Add 10+ revisions. Witness how the pagination now appears, and sits, at the bottom of the revision panel. Navigate through the results!
3. Select some revisions to preview them. Test that the "Apply" button works to apply revisions to the editor and close the revisions panel. (There shouldn't be any regression in the way folks can preview and apply revisions).
4. Make sure a "Reset" item is appended to the last page of the revisions results and not on any other page.
5. Make some changes to styles in the editor, but don't save. Now make sure an "Unsaved" item is prepended to the first page of the revisions results and not on any other page.

Because I've abstracted the pagination component from the Patterns page, check that there are no regressions between trunk and this PR:

| This PR  | Trunk |
| ------------- | ------------- |
| <img width="311" alt="Screenshot 2023-12-19 at 2 20 19 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/9761d7a6-eba9-4678-9f1e-056b8dc495b9">  | <img width="315" alt="Screenshot 2023-12-19 at 2 20 52 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/3c7373d9-c39c-4f7f-8739-cc94d90265f4">  |



### Testing Instructions for Keyboard
It should be possible to tab down the revision items and access the pagination navigation buttons.

<img width="283" alt="Screenshot 2023-12-19 at 2 18 49 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/57fe9807-8aab-4062-bf26-9d816c3e9e63">


## TODO
- [x] Style pagination component
- [x] Position pagination component, maybe in the sticky footer? Would require moving "Apply" button, perhaps inside every individual button?
- [x] Prevent full render when fetching uncached results (i.e., not already in store) - just need to remove condition renders on `shouldShowRevisions`
